Promote reusable combines through dedicated `flyd.combine` function

c-dante opened the pull request:
Currently, in a dependent stream you have to capture the parent stream in a closure:
```
var a = flyd.stream(0);
var b = a.map((x) => x * 2);
var dep = flyd.stream([a, b], (self, changed) => {
  // Required to defined b outside and capture it
  return a() + b();
});
```

Because it requires a closure, you have to write your combine functions in a non reusable fashion.

I'm proposing the following nonbreaking change to address this:
```
var sumCombine = (self, changed, x, y) => x() + y();

// ELSEWHERE
var a = flyd.stream([depA, depB], sumCombine);
var b = flyd.stream([depC, depD], sumCombine);
var allTheSums = flyd.stream([a, b], sumCombine);
```

This can be taken a step further for better, reusable join functions. However, it requires a breaking change:
```
flyd.stream([a, b, ...], (a, b, ..., self, changed) => {
});
```

Which allows reusable joins, nice partial functions, just all good things functional wants:
```
// Assuming functions are curried by default
var sumStreams = (a, b, c) => a() + b() + c();
var addFive = sum(flyd.stream(5));

var nice = flyd.stream([someDepA, someDepB], addFive);
```

Super sexy combines. Full access to `self` and `changed` if needed (but because return maps nicely, it's no worries)

c-dante commented:
I realized the test I wrote is a poor example (uses same variable names as the dependencies).

I can make a new pull request or write a more robust test after. something like the example in the request comment.

paldepind commented:
I think this is a very nice idea! :smile: For some use cases this is definitely more flexible and powerful. Thank you for sharing the idea and implementation.

I agree with you that the second approach is best. But it would require updating all the Flyd extension modules. Do you think it is worth it?

jplikesbikes commented:
I think this is worth the breaking change. I can immediately see places in my current code base that this would clean up and make simpler.

StreetStrider commented:
The first time I met flyd I was wondering why this haven't been done in the manner @c-dante offers :smile:

I was thinking that this is for sake of simplicity or that I'm dumb in some way. But looks like y'all realize this issue too. This change can open window for some hi-order stuff to apply, like Ramda and others.

ericgj commented:
I see the advantages, but it makes the dep function variadic, which makes it awkward to deal with, particularly with currying. You could do `flyd.stream([a,b,c], ([a,b,c],self,changed) => .... )` I suppose.

Personally I don't see the need for this change, as I tend to see defining dependent streams this way as kind of 'low-level', and don't see a problem with just passing in the stream(s) to a function that creates the dependent stream (basically the way all the modules work now).  I like the simple `(self, changed)` signature. But I've only used flyd for limited things so I'm sure I'm missing something.

c-dante commented:
@ericgj, the variadic nature of the dependent stream is captured in multiple places as it stands. Having the fixed-arity "self, changed" signature captures nothing about the stream combinator, which is why the current implementation looks to the closure to do comparisions, IE:
```
// a and b are streams
flyd.stream([a, b], (self, changed) => {
if (changed.indexOf(a) !== -1) // do something because `a` changed
if (changed.indexOf(b) !== -1) // do something because `b` changed
});
```
The problem with this signature is that you cannot curry the combinator, or even define it without a closure. My first change addresses the closure aspect in a non-breaking fashion by passing each dependency as the trailing variables. The problem with them being at the end is that you cannot curry this.

The array solution you suggest does nothing to address the problems outlined, and does not allow currying. Example:
```
var weirdArraySum = (aAndB) => aAndB[0] + aAndB[1];
flyd.stream([depA, depB], weirdArraySum);
```
But that isn't a curried function -- I cannot curry `weirdArraySum` since it forces a tuple to be applied. 

An alternative version with unnecessary boilerplate:
```
var sum = (a, b) => a + b;
flyd.stream([depA, depB], (deps, self, changed) => sum.apply(sum, deps));
```

ericgj commented:
I understand, I guess the feeling I am left with is, under what circumstances is something like this:

```javascript
var sum = (a, b, c) => a() + b() + c();
var sumStream = flyd.stream([streamA,streamB,streamC], sum);
```
more expressive/powerful than this:
```javascript
var lift = require('flyd/module/lift');  // or ramda lift would probably work too
var sum = (a,b,c) => a + b + c;
var sumStream = lift(sum, streamA, streamB, streamC);
```

c-dante commented:
So lifting and currying are not the same concept. You can solve the `sum` example by lifting, but that's not the issue we're addressing by putting the stream dependencies in the front of the combinator's signature.

Lifting is performing the operation of mapping from one space to another. This allows you to do things like write functions in the value space and then run those functions given promises (or streams). This is a completely different concern than currying.

 Currying is addressing functional composition and partial functions.

ericgj commented:
Two different issues -- one having to do with the combinator's signature being made variadic, the other that there are apparently ways of doing what you want to do already, namely with `lift`. 

But let me put it another way. You're suggesting a breaking change. I believe would affect not only the modules in this repo, but also some core functions like `map` and `ap`, as well as other people's modules out there in the wild. Nothing wrong with breaking changes, if they have a benefit and that benefit outweighs the cost of added complexity and cost of changing downstream code. But what I am asking is, can you give a real example of a case where it would be an added benefit to have this new API?  I assume you are doing something more complex than summing numbers off of streams :smile:

paldepind commented:
@c-dante

> Lifting is performing the operation of mapping from one space to another. This allows you to do things like write functions in the value space and then run those functions given promises (or streams). This is a completely different concern than currying.

It's not very relevant, but I don't think your terminology is accurate. I'm not sure what you mean with "space". But if I understand you correctly you differ between values and promises (or streams). But promises and streams are values as well. Lifting a function of one value (`map`) makes it possible to apply it to a functor. Lifting a function of more than one argument makes it possible to a apply it to a matching amount of applicative functors. I.e. lifting a function is applying it to a value inside another value.

@ericgj 

> Personally I don't see the need for this change, as I tend to see defining dependent streams this way as kind of 'low-level', and don't see a problem with just passing in the stream(s) to a function that creates the dependent stream (basically the way all the modules work now). I like the simple (self, changed) signature. But I've only used flyd for limited things so I'm sure I'm missing something.

I wrote something similar [the first time](https://github.com/paldepind/flyd/issues/3#issuecomment-93934808) this issue was brought up. But my understanding now is that people like to use the versatility of `stream` in normal code. And I think that is valid.

In hindsight I regret making `stream(initalValue)` and `stream(deps, body)` the same function. The second should probably have been name `combine`. If we do a breaking change I think the function should be split in two at the same time.

> I see the advantages, but it makes the dep function variadic, which makes it awkward to deal with, particularly with currying. You could do flyd.stream([a,b,c], ([a,b,c],self,changed) => .... ) I suppose.

I don't understand what will be made more awkward? The dep function will be applied with varying numbers of arguments. But the amount is predictable to the user so I think it would work with currying?

StreetStrider commented:
@paldepind 
>  The dep function will be applied with varying numbers of arguments. But the amount is predictable to the user so I think it would work with currying?

Exactly. I was thinking the same way. So, if we want to sum *two* streams, any *binary* function will work for it, like `Ramda.add` or others. So «variadicity» here will be predictable and is completely up to user. User defines arity of function by passing certain number of dependencies and supplies an appropriate function.

c-dante commented:
Yeah, that's what I was trying to capture without using the words "category", "morphism", or "functor". Although I think there's a subtlety between the idea of lift; given `f : X -> Y` and `g : Z -> Y`, lifting `f` through `Z` would be a function `h` such that `h o g = f` -- that is, a function that maps `X -> Y` though `Z`

On topic:
I absolutely agree with you that `stream(deps, body)` constructor is misleading. How do you feel about deprecating that signature and adding in `flyd.combine([a, b, ...], (a, b, ..., self, changed))`?

Similarly, `a.combine([b, c, ...], (a, b, c, ..., self, changed))`.

That sounds like a nice way to address moving forward with the library, having a nicely curried string combine, and being clear about usage?

ericgj commented:
> In hindsight I regret making stream(initalValue) and stream(deps, body) the same function. The second should probably have been name combine. If we do a breaking change I think the function should be split in two at the same time.

Definitely agree.

>  But my understanding now is that people like to use the versatility of stream in normal code. And I think that is valid.

In my opinion, it's _nice_ to have to capture the streams in a closure: it discourages stream getter calls spreading all over your code. But I understand people have different preferences...

> I don't understand what will be made more awkward? The dep function will be applied with varying numbers of arguments. But the amount is predictable to the user so I think it would work with currying?

Yes, it's probably not a practical issue. I think the more awkward case is if you have a dep function that can apply to a variable number of streams, and you need to use `self` or  `changed`. You have to slice them out from the right side of the arguments. Less of a big deal with ES6 of course.

But doesn't really matter to me, esp as there are other reasons to change the interface anyway, and I wouldn't normally use this interface anyway outside of modules.

paldepind commented:
@c-dante

> Although I think there's a subtlety between the idea of lift; given `f : X -> Y` and `g : Z -> Y`, lifting `f` through `Z` would be a function `h` such that `h o g = f` -- that is, a function that maps `X -> Y` though `Z`

Indeed. The "apply to value inside container" intuition is harder to apply in that case. It can be done. I like to think of a function as a value with a hole. You need to fill in the hole with another value to get the final value. Thus function is a container of an "unfinished" value and you can lift a function to apply to the unfinished value.

> I absolutely agree with you that `stream(deps, body)` constructor is misleading. How do you feel about deprecating that signature and adding in `flyd.combine([a, b, ...], (a, b, ..., self, changed))`?

I think that is a great idea. If we break backwards compatibility we should grab the chance an make other changes as well. This will also make it more worthwhile.

Would you like to update the pull request?

c-dante commented:
Sounds good -- I'll implement a `flyd.combine([a, b, ...], (a, b, ..., self, changed))` and update the tests to match.

Do you have any other breaking changes?

Is accessing `stream.val` or `stream.deps` (or others) a part of the public interface?

paldepind commented:
> Sounds good -- I'll implement a flyd.combine([a, b, ...], (a, b, ..., self, changed)) and update the tests to match.

That sounds really good :)

> Do you have any other breaking changes?

Nope.

> Is accessing stream.val or stream.deps (or others) a part of the public interface?

Some modules use `stream.val`. But I don't think there is a reason to change that? Is there?

Let me know, if you have any questions.

StreetStrider commented:
Is `combine` name ok? Maybe someone has a better name?

Is it possible, after `flyd.stream(deps, body)` deprecation, after a couple of releases *reuse* it with semantics of `combine`, and then deprecate `combine`? :smile: I really like `flyd.stream` name.

c-dante commented:
Updated the pull request to match what I'm offering now.

* Deprecated `flyd.stream([a, b, ...], (self, changed) => {})` in favor of `flyd.combine([a, b, ....], (a, b, ..., self, changed) => {})`
* Replaced all library and test usage of dependent streams to `flyd.combine`
* Updated all tests to use `flyd.combine`
* Included tests for deprecated code (duplicate of atomic update tests) for coverage

Let me know how you test the modules folder. I'm going to run through and update them anyway, though.

paldepind commented:
> Is it possible, after flyd.stream(deps, body) deprecation, after a couple of releases reuse it with semantics of combine, and then deprecate combine? :smile: I really like flyd.stream name.

To me the whole point is that it should be two separate functions because they do two separate things.

@c-dante That sounds very good! I'll look over it as soon as I get the time.

> Let me know how you test the modules folder.

The only way is to run `mocha` in each module directory. Actually. I've just realized that the tests for the modules are broken after they got moved into the main repository. All the individual module files has been renamed to `index.js` in their directory but the filename used in the tests has not been updated.

Should I fix it so that you can rebase on that?

c-dante commented:
Yeah if you don't mind. Shouldn't be hard to rip through the tests and update them.

paldepind commented:
The PR looks very good! I don't really think leaving in the deprecated code is necessary though?

I think having dependencies last makes sense. It's also more consistent with most Flyd modules that takes the function as their first argument.

c-dante commented:
Okay, I'll do the following, then we should be good to merge the PR:
* Update to be consistent with the rest of the library: `combine(fn, deps)`
* Update main test to match signature
* Update all current `flyd/module/*` to refer to the main `flyd.js` and update their tests
* Update documentation to reflect these changes

I'm fine axing the deprecated code, let me know if you'd like me to do that.

paldepind commented:
@c-dante That sounds like a very good plan.

I don't think deprecation is worth it for such a small change. But I don't have strong feelings so feel free to leave it in og you prefer.

I've just updated the tests for the modules so that they point to their correctly named files. I've mistakenly made some of them refer to the Flyd browser build in the root though.

jplikesbikes commented on lib/index.js:
```diff
@@ -255,9 +262,30 @@ function stream(arg, fn) {
 var transduce = curryN(2, function(xform, source) {
   xform = xform(new StreamTransformer());
-  return stream([source], function(self) {
+  return combine([source], function(self) {
```
I believe this should be
```
return combine([source], function(source, self) {
```

c-dante replied:
Yeah, good catch.

When I address the changelist I posted I'll hit this as well as add more tests for the transduce code.

c-dante commented:
This sums up the diff for the pull request so far:
* Deprecate `flyd.stream([a, b, ...], (self, changed))` in favor of a curried `flyd.combine((a, b, ..., self, changed), [a, b, ...])`
 * Update `README.md` to reflect this with regard to dependent streams
 * Updated tests to reflect this
 * Update the `sum` example with `flyd.combine`
* Internal use of `combine` uses the `self` argument to update streams -- this preserves `undefined` value events. (`scan` and `merge`)
 * Added test to assert this
* Consistency with using `stream.val` for internal functions -- This should go one way or the other (all in on `.val` or always using the getter call)
* Updated all modules to `flyd.combine` and passing tests
 * `droprepeats` module: Removed `chai` reference in favor of node's `assert`
* Rewrote `streamMerge` module. Now passes old tests (current flyd master does not)
* Added `bower_components` to `.gitignore`
* Fixed the typo @jplikesbikes pointed out in my `flyd.transduce`

Let me know what you think, and if there's anything outstanding.

paldepind commented:
Thank you a lot for doing this.

paldepind commented:
I've published a new version and published it to npm. Great work @c-dante
